# P1: Analyze Employee Salary Data using Pandas

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('employee_earnings_report_2024.csv')

df.columns = df.columns.str.replace(' ', '_')

numeric_cols = ['REGULAR', 'RETRO', 'OTHER', 'OVERTIME', 'INJURED', 'DETAIL', 'QUINN_EDUCATION', 'TOTAL_GROSS']

for col in numeric_cols:
    df[col] = pd.to_numeric(df[col].astype(str).str.replace(',', ''), errors='coerce')

df[numeric_cols] = df[numeric_cols].fillna(0)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 573 entries, 0 to 572
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   NAME             573 non-null    object 
 1   DEPARTMENT_NAME  573 non-null    object 
 2   TITLE            573 non-null    object 
 3   REGULAR          573 non-null    float64
 4   RETRO            573 non-null    float64
 5   OTHER            573 non-null    float64
 6   OVERTIME         573 non-null    float64
 7   INJURED          573 non-null    float64
 8   DETAIL           573 non-null    float64
 9   QUINN_EDUCATION  573 non-null    float64
 10  TOTAL_GROSS      573 non-null    float64
 11  POSTAL           572 non-null    float64
dtypes: float64(9), object(3)
memory usage: 53.8+ KB


### 1. How many rows and columns does the dataset have?

In [2]:
rows, columns = df.shape
print(f"Number of rows: {rows}")
print(f"Number of columns: {columns}")

Number of rows: 573
Number of columns: 12


**Interpretation:** The dataset contains 573 employee records, with 12 different attributes recorded for each employee.

### 2. Create an EMP1 DataFrame containing only the variables Employee Name, Department, and Total Earnings.

In [3]:
EMP1 = df[['NAME', 'DEPARTMENT_NAME', 'TOTAL_GROSS']]
print(EMP1.head())

                 NAME   DEPARTMENT_NAME  TOTAL_GROSS
0  Hallahan,Timothy J  Traffic Division    240146.89
1   Hesford,Edward D.  Traffic Division    227972.68
2       Aleman,Wilson  Traffic Division    198690.08
3       Bloom,David M           Finance    194429.06
4    Vilar,Alfredo J.  Traffic Division    182660.81


**Interpretation:** The EMP1 DataFrame has been created to provide a focused view on employee names, their departments, and total earnings.

### 3. Create EMP2 dataframe containing all variables of the dataset except Overtime Earnings.

In [4]:
EMP2 = df.drop(columns=['OVERTIME'])
print(EMP2.head())

                 NAME   DEPARTMENT_NAME                          TITLE  \
0  Hallahan,Timothy J  Traffic Division  Supv-Parking Meter Operations   
1   Hesford,Edward D.  Traffic Division     Supvising Traffic Engineer   
2       Aleman,Wilson  Traffic Division          Asst Supv Traffic Eng   
3       Bloom,David M           Finance     Chief Financial Officer(F)   
4    Vilar,Alfredo J.  Traffic Division            Sr Traffic Engineer   

     REGULAR    RETRO    OTHER  INJURED  DETAIL  QUINN_EDUCATION  TOTAL_GROSS  \
0  114623.09     0.00     0.00      0.0     0.0              0.0    240146.89   
1  133169.08     0.00     0.00      0.0     0.0              0.0    227972.68   
2  111273.88  2034.51   984.78      0.0     0.0              0.0    198690.08   
3  194429.06     0.00     0.00      0.0     0.0              0.0    194429.06   
4  107792.88  1869.34  2153.71      0.0     0.0              0.0    182660.81   

   POSTAL  
0  1940.0  
1  2131.0  
2  2191.0  
3  2132.0  
4  2155.

**Interpretation:** The EMP2 DataFrame contains all employee data except for their overtime earnings, allowing for analysis without its influence.

### 4. Select employees with Total Earnings > 100,000 and Overtime Earnings > 5,000. How many such employees exist?

In [5]:
high_earners = df[(df['TOTAL_GROSS'] > 100000) & (df['OVERTIME'] > 5000)]
num_employees = len(high_earners)
print(high_earners.head())
print(f"Number of employees with Total Earnings > $100,000 and Overtime > $5,000: {num_employees}")

                 NAME   DEPARTMENT_NAME                          TITLE  \
0  Hallahan,Timothy J  Traffic Division  Supv-Parking Meter Operations   
1   Hesford,Edward D.  Traffic Division     Supvising Traffic Engineer   
2       Aleman,Wilson  Traffic Division          Asst Supv Traffic Eng   
4    Vilar,Alfredo J.  Traffic Division            Sr Traffic Engineer   
6      Smith,Joseph M  Traffic Division       Supv-Parking Enforcement   

     REGULAR    RETRO    OTHER   OVERTIME  INJURED  DETAIL  QUINN_EDUCATION  \
0  114623.09     0.00     0.00  125523.80      0.0     0.0              0.0   
1  133169.08     0.00     0.00   94803.60      0.0     0.0              0.0   
2  111273.88  2034.51   984.78   84396.91      0.0     0.0              0.0   
4  107792.88  1869.34  2153.71   70844.88      0.0     0.0              0.0   
6  116182.87     0.00  4393.42   52403.89      0.0     0.0              0.0   

   TOTAL_GROSS  POSTAL  
0    240146.89  1940.0  
1    227972.68  2131.0  
2    

**Interpretation:** There are 58 employees who have earned more than \$100,000 in total with over \$5,000 in overtime pay.

### 5. Create an EMP3 DataFrame containing only employees from the Finance Department.

In [6]:
EMP3 = df[df['DEPARTMENT_NAME'] == 'Finance']
print(EMP3.head())

                       NAME DEPARTMENT_NAME                           TITLE  \
3             Bloom,David M         Finance      Chief Financial Officer(F)   
16            Macrae,Yvonne         Finance            Director (NonAcad) D   
22          Racanelli,James         Finance            Director (NonAcad) D   
25           Beltre,Cynthia         Finance            Director (NonAcad) D   
26  Dawkins,Blair Elizabeth         Finance  Deputy Chief Financial Off (E)   

      REGULAR  RETRO   OTHER  OVERTIME  INJURED  DETAIL  QUINN_EDUCATION  \
3   194429.06    0.0     0.0       0.0      0.0     0.0              0.0   
16  150143.76    0.0  1500.0       0.0      0.0     0.0              0.0   
22  145479.62    0.0     0.0       0.0      0.0     0.0              0.0   
25  142536.72    0.0     0.0       0.0      0.0     0.0              0.0   
26  141921.13    0.0     0.0       0.0      0.0     0.0              0.0   

    TOTAL_GROSS  POSTAL  
3     194429.06  2132.0  
16    151643.76 

**Interpretation:** The EMP3 DataFrame isolates all records for employees working exclusively in the Finance department for targeted analysis.

### 6. Create an EMP4 DataFrame sorted in descending order of Total Earnings. Display the first 10 rows.

In [7]:
EMP4 = df.sort_values(by='TOTAL_GROSS', ascending=False)
print(EMP4.head(10))

                 NAME   DEPARTMENT_NAME                          TITLE  \
0  Hallahan,Timothy J  Traffic Division  Supv-Parking Meter Operations   
1   Hesford,Edward D.  Traffic Division     Supvising Traffic Engineer   
2       Aleman,Wilson  Traffic Division          Asst Supv Traffic Eng   
3       Bloom,David M           Finance     Chief Financial Officer(F)   
4    Vilar,Alfredo J.  Traffic Division            Sr Traffic Engineer   
5     Dimarzo,Dominic  Traffic Division            Traffic Signal Supv   
6      Smith,Joseph M  Traffic Division       Supv-Parking Enforcement   
7     Gove,Nicholas S  Traffic Division                   Deputy Chief   
8     Richards,Jay J.  Traffic Division  Asst Supv-Parking Enforcement   
9    D'Avolio,Kristen   Adult Education    Senior Director (BASAS 12B)   

     REGULAR    RETRO    OTHER   OVERTIME    INJURED  DETAIL  QUINN_EDUCATION  \
0  114623.09     0.00     0.00  125523.80       0.00     0.0              0.0   
1  133169.08     0.00  

**Interpretation:** The EMP4 DataFrame lists employees from highest to lowest total earnings, with the top 10 highest earners displayed.

### 7. Add a new column, Annual_Bonus, in EMP2 dataframe using the formula Annual_Bonus = 10% of Regular Earnings for each employee.

In [8]:
EMP2['Annual_Bonus'] = EMP2['REGULAR'] * 0.10
print(EMP2[['NAME', 'DEPARTMENT_NAME', 'TITLE', 'REGULAR', 'TOTAL_GROSS', 'Annual_Bonus']].head())

                 NAME   DEPARTMENT_NAME                          TITLE  \
0  Hallahan,Timothy J  Traffic Division  Supv-Parking Meter Operations   
1   Hesford,Edward D.  Traffic Division     Supvising Traffic Engineer   
2       Aleman,Wilson  Traffic Division          Asst Supv Traffic Eng   
3       Bloom,David M           Finance     Chief Financial Officer(F)   
4    Vilar,Alfredo J.  Traffic Division            Sr Traffic Engineer   

     REGULAR  TOTAL_GROSS  Annual_Bonus  
0  114623.09    240146.89     11462.309  
1  133169.08    227972.68     13316.908  
2  111273.88    198690.08     11127.388  
3  194429.06    194429.06     19442.906  
4  107792.88    182660.81     10779.288  


**Interpretation:** An 'Annual_Bonus' column, calculated as 10% of regular earnings, has been added to the EMP2 DataFrame.

### 8. Calculate the average Regular Earnings and Overtime Earnings for: (i) the entire dataset (ii) each Department.

In [9]:
# (i) Average for the entire dataset
average_all = df[['REGULAR', 'OVERTIME']].mean()
print("Average Earnings for the Entire Dataset:")
print(average_all)

# (ii) Average for each department
average_by_dept = df.groupby('DEPARTMENT_NAME')[['REGULAR', 'OVERTIME']].mean()
print("\nAverage Earnings by Department:")
print(average_by_dept)

Average Earnings for the Entire Dataset:
REGULAR     53451.236545
OVERTIME     6607.644241
dtype: float64

Average Earnings by Department:
                          REGULAR      OVERTIME
DEPARTMENT_NAME                                
Adult Education      77902.152121      0.000000
Athletics            14888.902843      0.000000
Auditing Department  83934.880976   2025.609756
Finance              94123.588065      0.000000
Traffic Division     55133.765246  10117.841940


**Interpretation:** The analysis shows the overall and department-specific average earnings based on the available 573 records.

### 9. Find the top 5 highest-paid employees in the entire dataset.

In [10]:
top_5_paid = df.nlargest(5, 'TOTAL_GROSS')
print(top_5_paid[['NAME', 'DEPARTMENT_NAME', 'TITLE', 'TOTAL_GROSS']])

                 NAME   DEPARTMENT_NAME                          TITLE  \
0  Hallahan,Timothy J  Traffic Division  Supv-Parking Meter Operations   
1   Hesford,Edward D.  Traffic Division     Supvising Traffic Engineer   
2       Aleman,Wilson  Traffic Division          Asst Supv Traffic Eng   
3       Bloom,David M           Finance     Chief Financial Officer(F)   
4    Vilar,Alfredo J.  Traffic Division            Sr Traffic Engineer   

   TOTAL_GROSS  
0    240146.89  
1    227972.68  
2    198690.08  
3    194429.06  
4    182660.81  


**Interpretation:** The top 5 highest-paid employees in the dataset are listed, showing the peak compensation levels.

### 10. Display the employee(s) with the lowest Total Earnings in each department.

In [11]:
lowest_earners_idx = df.groupby('DEPARTMENT_NAME')['TOTAL_GROSS'].idxmin()
lowest_earners = df.loc[lowest_earners_idx]
print(lowest_earners[['NAME', 'DEPARTMENT_NAME', 'TITLE', 'TOTAL_GROSS']])

                      NAME      DEPARTMENT_NAME                     TITLE  \
518       Nguyen,Kimvy Thi      Adult Education        Adult Ed-Tchr/Asst   
563         Molle,Kimberly            Athletics            Coach (NonTPP)   
405  Morrison,Titus Eugene  Auditing Department             Sr Accountant   
572  Kennedy Sr.,Donald R.              Finance      Employment Agreement   
571          Floyd,Chloe L     Traffic Division  Parking Meter Supervisor   

     TOTAL_GROSS  
518      4309.17  
563      1028.61  
405     21399.94  
572       112.50  
571       128.60  


**Interpretation:** The lowest-earning employee from each department has been identified, highlighting the minimum pay levels in the dataset.

### 11. Which department has the highest average Overtime Earnings?

In [12]:
avg_overtime_by_dept = df.groupby('DEPARTMENT_NAME')['OVERTIME'].mean()
top_overtime_dept = avg_overtime_by_dept.idxmax()
print(f"Department with the highest average Overtime Earnings: {top_overtime_dept}")

Department with the highest average Overtime Earnings: Traffic Division


**Interpretation:** The Traffic Division has the highest average overtime earnings, indicating a greater reliance on overtime work in that department.

### 12. Add a new column, Level, in EMP2 dataframe in terms of three earning categories
- **Low:** Total Earnings < 50,000 [cite: 20]
- **Medium:** 50,000 ≤ Total Earnings < 100,000 [cite: 21]
- **High:** Total Earnings ≥ 100,000 [cite: 22]

In [13]:
bins = [-np.inf, 50000, 100000, np.inf]
labels = ['Low', 'Medium', 'High']

EMP2['Level'] = pd.cut(EMP2['TOTAL_GROSS'], bins=bins, labels=labels, right=False)

print(EMP2[['NAME', 'TOTAL_GROSS', 'Level']].head())

print("\nValue counts for each earning level:")
print(EMP2['Level'].value_counts())

                 NAME  TOTAL_GROSS Level
0  Hallahan,Timothy J    240146.89  High
1   Hesford,Edward D.    227972.68  High
2       Aleman,Wilson    198690.08  High
3       Bloom,David M    194429.06  High
4    Vilar,Alfredo J.    182660.81  High

Value counts for each earning level:
Level
Medium    231
Low       225
High      117
Name: count, dtype: int64


**Interpretation:** Employees have been categorized into 'Low' (225), 'Medium' (231), and 'High' (117) earning levels for easier segmentation.